[change] Move python dependencies to requirements.txt file

shwetd19 opened the pull request:
## Checklist

- [x] I have read the [[OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html)](http://openwisp.io/docs/developer/contributing.html).
- [x] I have tested the changes in this pull request to make sure they work.
- [ ] I have added or updated tests for any new or changed code.
- [ ] I have updated the documentation where needed.

---

## Reference to Existing Issue

Closes #52.

---

## Description of Changes

This pull request moves Python dependencies from `pip.yml` to a `requirements.txt` file for easier management. Here’s what was done:

1. **Created a `requirements.txt` file**:
   - Added all dependencies, such as `Flask`, `Werkzeug`, and `uwsgi`, into this file to keep things organized.

2. **Updated `pip.yml`**:
   - Adjusted the Ansible tasks to install dependencies from `requirements.txt` instead of listing them directly in the file.
   - Added steps to:
     - Update `pip` and related tools like `setuptools` and `wheel`.
     - Install dependencies in a virtual environment using the `requirements.txt` file.

3. **Dependabot Integration**:
   - Made sure tools like Dependabot can track and update dependencies in `requirements.txt` automatically.

4. **Tested the Changes**:
   - Ran the updated playbook to confirm everything works as expected.

These changes make it easier to manage and update dependencies while keeping the setup clean and consistent.

pandafy commented on tasks/pip.yml:
```diff
@@ -20,13 +14,14 @@
-- name: Install Flask, Werkzeug and uWSGI
+- name: Read local requirements.txt
+  local_action: command cat requirements.txt
+  register: requirements
```
Are you sure this is correct? Did you try running the tests locally?
https://github.com/openwisp/ansible-wireguard-openwisp?tab=readme-ov-file#how-to-run-tests

shwetd19 commented:
@pandafy  , I've fixed the path for requirements.txt file, now it's working.. can you please check nd lmk, Thanks!

nemesifier commented on tasks/pip.yml:
```diff
@@ -1,14 +1,8 @@
-      - wheel
-      - attrs
-      - importlib-metadata
-      - packaging
```
I wouldn't move these because we're always upgrade to the latest version available.

The aim is to move only the python packages that we are pinning to as pecific version range, so that dependabot can help us update those more with less intervention from our end.

nemesifier reviewed (changes requested):
See my comment below, please also take a look at the message returned by the QA checks failure regarding the lack of capital letter after the square bracket, eg:

`[prefix] done something` should be `[prefix] Done something`.

shwetd19 commented:
Hey @pandafy @nemesifier I've modified the changes can you please review this once. Thanks!


- Removed unnecessary dependencies (`setuptools`, `wheel`, `attrs`, `importlib-metadata`, `packaging`) from `requirements.txt`
- Updated `pip.yml`:
  - Ensured `setuptools`, `wheel`, `attrs`, `importlib-metadata`, and `packaging` are installed directly via pip task
  - Replaced reading `requirements.txt` via `local_action` with explicit file copy to virtualenv
  - Changed package installation to use the copied `requirements.txt` file
  - Added condition to install packages only if `requirements.txt` changes

nemesifier commented on tasks/pip.yml:
```diff
@@ -1,5 +1,4 @@
 ---
-
```
Don't remove this line.

nemesifier commented on tasks/pip.yml:
```diff
@@ -20,19 +19,22 @@
   retries: 5
   delay: 10
-  register: result
+  register: result 
```
```suggestion
  register: result
```

nemesifier commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Flask~=3.0.3
+uwsgi~=2.0.26
```
```suggestion
uwsgi~=2.0.28
```

nemesifier reviewed (changes requested):
See my comments below.
Again I want to bring to your attention to please also take a look at the message returned by the QA checks failure regarding the lack of capital letter after the square bracket, eg:

`[prefix] done something` should be `[prefix] Done something`.

shwetd19 commented:
@nemesifier I've made the suggested changes, please let me know if any changes are required, also pls forgive me if I'm making silly mistakes here, it's just I'm testing everything I can locally..

nemesifier commented on tasks/pip.yml:
```diff
@@ -20,19 +20,22 @@
   retries: 5
   delay: 10
-  register: result
+  register: result 
```
```suggestion
  register: result
```

nemesifier commented on tasks/pip.yml:
```diff
@@ -20,19 +20,22 @@
+    dest: "{{ virtualenv_path }}/requirements.txt"
+  register: requirements_copy
+
+- name: Install Python packages from Requirements.txt
```
```suggestion
- name: Install Python dependencies
```

nemesifier commented on tasks/pip.yml:
```diff
@@ -20,19 +20,22 @@
   notify: reload supervisor
 
-- name: Install Flask, Werkzeug and uWSGI
+- name: Copy Requirements.txt to virtualenv
```
```suggestion
- name: Copy requirements.txt to virtualenv dir
```

nemesifier commented on tasks/pip.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 
-- name: Update pip & related tools
+- name: Update Pip & related tools
```
```suggestion
- name: Update pip & related tools
```

nemesifier reviewed (changes requested):
Small adjustments for consistency

shwetd19 commented:
@nemesifier Thanks you very much for this! I'll try to learn as much as I can for never miss an oversight like this

pandafy commented on requirements.txt:
This file should be inside the files/ directory.

nemesifier replied:
Good point, thanks!

shwetd19 replied:
Moved `requirements.txt` from the root directory to `files/requirements.txt`

pandafy commented on tasks/pip.yml:
```diff
@@ -20,13 +20,15 @@
   notify: reload supervisor
 
-- name: Install Flask, Werkzeug and uWSGI
+- name: Copy requirements.txt to virtualenv dir
+  copy:
+    src: "../requirements.txt"
+    dest: "{{ virtualenv_path }}/requirements.txt"
+  register: requirements_copy
```
I am unsure about the destination of the requirements file. With the current changes, it will be stored at 
`/opt/wireguard-openwisp/env/requirements.txt`. It will be unusual to store the requirements file inside the virtual environment. 

```suggestion
- name: Copy requirements.txt 
  copy:
    src: requirements.txt
    dest: "{{ openwisp2_wireguard_path }}/requirements.txt"
```

pandafy replied:
`requirements.txt` is very command in the Python eco-system, so we can keep using that. 

With my suggested changes, the file will be copied to `/opt/wireguard-openwisp/requirements.txt`

nemesifier replied:
Ok, thanks for clarifying.

shwetd19 replied:
Updated the pip install task to use the requirements file from the new location

pandafy commented on tasks/pip.yml:
```diff
@@ -36,3 +38,4 @@
   delay: 10
   register: result
   until: result is success
+  when: requirements_copy is changed
```
This should be run always. 
```suggestion
```

pandafy reviewed (changes requested):
I took a closer look and found the following area for improvements.

nemesifier commented on tasks/pip.yml:
```diff
@@ -1,3 +1,4 @@
+# tasks/pip.yml
```
This addition is irrelevant so let's remove it.
```suggestion
```

nemesifier commented on tasks/pip.yml:
```diff
@@ -35,4 +38,4 @@
   delay: 10
   register: result
-  until: result is success
+  until: result is success
```
Configure your editor to always leave a blank line at the end of the file please.

```suggestion
  until: result is success

```

shwetd19 commented:
Hey @nemesifier @pandafy both suggested fixes are done! pls can you check once

pandafy reviewed (approved):
LGTM! 

Thanks for contributing @shwetd19 🚀
